lib/_http_agent.js

```diff
@@ -268,6 +272,24 @@ function installListeners(agent, s, options) {
   }
   s.on('close', onClose);
 
+  function onTimeout() {
+    debug('CLIENT socket onTimeout after', s.timeout, 'ms');
+    const name = agent.getName(options);
+    // If free socket timeout, must remove it from freeSockets list immediately
+    // to prevent new requests from being sent through this timeout socket.
+    if (agent.freeSockets[name] && agent.freeSockets[name].indexOf(s) !== -1) {
+      s.destroy();
+      agent.removeSocket(s, options);
+      debug('CLIENT free socket destroy');
+    } else if (s.listeners('timeout').length === 1) {
+      // No req timeout handler, agent must destroy the socket.
```

Comment:
Isn't this inconsistent with the "non-agent" flow, where no timeout listener is just a noop?

If the socket is not in the free list then I believe the current behaviour is to let the user choose to handle it (or not).

I don't think this branch is necessary here (at least not in the context of this PR).

Reply:
i.e. a test like this might be relevant:

```js
const http = require('http');
const common = require('../common');

function test (agent) {
  const options = {
    method: 'GET',
    port: undefined,
    host: '127.0.0.1',
    path: '/',
    agent
  };
  
  const server = http.createServer(function(req, res) {
    // This space intentionally left blank
  });
  
  server.listen(0, options.host, function() {
    options.port = this.address().port;
    const req = http.request(options, function(res) {
      // This space intentionally left blank
    });
    req.on('close', function() {
      server.close();
    });
    const s = req.setTimeout(1);
    req.on('socket', common.mustCall((socket) => {
      socket.on('timeout', common.mustCall());
      socket.on('error', common.mustNotCall());
    }));
    req.on('timeout', common.mustCall());
    req.on('error', common.mustNotCall());
    req.end();
  });
}

test();
test(new http.Agent({ timeout: 100 }));
```

Reply:
This timeout socket is generated by our agent. We need to actively notify the user of the `error` event, otherwise the user's code callback will not be executed because they does not handle the timeout event logic.

Reply:
Exactly. That's the way it's supposed to work. There is no `'error'` event here, just a `'timeout'` event. See, https://nodejs.org/dist/latest-v13.x/docs/api/http.html#http_event_timeout.

> Emitted when the underlying socket times out from inactivity. **This only notifies that the socket has been idle. The request must be aborted manually.**

Reply:
@fengmk2: Any update here? I'd like it if we could fix the timeout of free list sockets that this PR addresses.